### PR TITLE
[SR-4363] Add SM3 Cryptographic Hash Algorithm

### PR DESCRIPTION
**Pull request by Pslydhh**
- Add sm3 expr for get SM3 hash value for any string as 8 words as hexadecimal seperate by ' '.
- Implement SM3 Cryptographic Hash Algorithm
http://www.oscca.gov.cn/sca/xxgk/2010-12/17/content_1002389.shtml
https://cloud.tencent.com/developer/article/1634816
https://blog.csdn.net/a344288106/article/details/80094878
https://github.com/guanzhi/GmSSL
https://github.com/NEWPLAN/SMx

### CODE REVIEW

File: be/src/util/sm3.cpp [hunk 1]
```diff
@@ -0,0 +1,241 @@
+// This file is licensed under the Elastic License 2.0. Copyright 2021 StarRocks Limited.
```
Comment (dirtysalt) [inline be/src/util/sm3.cpp:1]: Is sm3 public domain ? or has any license?

File: be/src/exprs/vectorized/string_functions.cpp [hunk 2]
```diff
@@ -2052,6 +2053,41 @@ ColumnPtr StringFunctions::unhex(FunctionContext* context, const starrocks::vect
...
+    unsigned long message_len = input_str.size;
+    if (message_len > 0) {
+        unsigned char output[32];
+
+        // output as 256 bits(32 bytes) result.
+        Sm3::sm3_compute(message, message_len, output);
+        result << std::hex << std::setfill('0');
+
+        // first 4 bytes;
+        result << std::setw(2) << (output[0] & 0xFF);
```
Comment (dirtysalt) [inline be/src/exprs/vectorized/string_functions.cpp:2070]: could this piece code be integrated into the following loop?

```
if ((i % 4) == 0 && i != 0) {
}
```

Comment (Pslydhh) [reply be/src/exprs/vectorized/string_functions.cpp:2070]: ok

File: be/test/exprs/vectorized/string_fn_sm3_test.cpp [hunk 3]
```diff
@@ -0,0 +1,41 @@
+// This file is licensed under the Elastic License 2.0. Copyright 2021 StarRocks Limited.
+#include <gtest/gtest.h>
+
+#include "exprs/vectorized/string_functions.h"
+
+namespace starrocks {
+namespace vectorized {
+
+class StringFunctionSm3Test : public ::testing::Test {};
+TEST_F(StringFunctionSm3Test, abcA1Test) {
```
Comment (kangkaisen) [inline be/test/exprs/vectorized/string_fn_sm3_test.cpp:10]: Need more test cases, const ,nullable,  NULL Literal

File: be/src/util/sm3.cpp (hunk 1 above)
Comment (chaoyli) [reply be/src/util/sm3.cpp:1]: different sm3 implementation have there own's copyright. If we use it, we should keep the license as same as the source. And also add license to notice.txt.

Comment (Pslydhh) [reply be/src/util/sm3.cpp:1]: > Is sm3 public domain ? or has any license?

SM3 is a cryptographic hash function standard adopted by the government of the people's Republic of China.

**Review by kangkaisen — APPROVED** [review]

**Review by dirtysalt — APPROVED** [review]

**Pull request merged**